GTG/backends/backend_caldav.py

```diff
@@ -0,0 +1,1037 @@
+                    continue
+                parents = PARENT_FIELD.get_dav(todo)
+                if (not parents  # no parent mean no relationship on build
+                        or parents[0] in known_todos  # already known parent
+                        or self.datastore.get_task(uid)):  # already known uid
+                    yield todo
+                    known_todos.add(uid)
+            if loop >= MAX_CALENDAR_DEPTH:
+                logger.error("Too deep, %dth recursion isn't allowed", loop)
+                break
+
+    def _get_calendar_tasks(self, calendar: iCalendar):
+        """Getting all tasks that has the calendar tag"""
+        for uid in self.datastore.get_all_tasks():
+            task = self.datastore.get_task(uid)
+            if CATEGORIES.has_calendar_tag(task, calendar):
+                yield uid, task
+
+    #
+    # Utility methods
+    #
+
+    def _get_todo_and_calendar(self, task: Task):
+        """For a given task, try to get the todo out of the cache and figures
+        out its calendar if one is linked to it"""
+        todo, calendar = self._cache.get_todo(UID_FIELD.get_gtg(task)), None
+        # lookup by task
+        for __, calendar in self._cache.calendars:
+            if CATEGORIES.has_calendar_tag(task, calendar):
+                logger.debug('Found from task tag %r and %r', todo, calendar)
+                return todo, calendar
+        cname = task.get_attribute('calendar_name', namespace=self.namespace)
+        curl = task.get_attribute("calendar_url", namespace=self.namespace)
+        if curl or cname:
+            calendar = self._cache.get_calendar(name=cname, url=curl)
+            if calendar:
+                logger.debug('Found from task attr %r and %r', todo, calendar)
+                return todo, calendar
+        if todo and getattr(todo, 'parent', None):
+            logger.debug('Found from todo %r and %r', todo, todo.parent)
+            return todo, todo.parent
+        return None, None
+
+    @property
+    def namespace(self):
+        return "caldav:%s" % self._parameters['service-url']
+
+
+class Field:
+    """ Basic field representation.
+
+    Allows to extract neutral values from GTG Task (attributes in integer or
+    tags without '@' for example) and from vTodo (translated datetime).
+    """
+
+    def __init__(self, dav_name: str,
+                 task_get_func_name: str, task_set_func_name: str,
+                 ignored_values: list = None):
+        self.dav_name = dav_name
+        self.task_get_func_name = task_get_func_name
+        self.task_set_func_name = task_set_func_name
+        self.ignored_values = ignored_values or ['', 'None', None]
+
+    def _is_value_allowed(self, value):
+        return value not in self.ignored_values
+
+    def get_gtg(self, task: Task, namespace: str = None):
+        "Extract value from GTG.core.task.Task according to specified getter"
+        return getattr(task, self.task_get_func_name)()
```

Comment:
Why not call the getter straight from the task?

Reply:
I'm not sure what you mean. Could you give an example ?
The `Field` class is suppose to be a one attribute translator made destined to get and set attribute from task. Getters and setters are set up in `task_get_func_name` and `task_set_func_name` because most of the current `Task` API is made through getters and setters and no property. But that's a part I'm going to alter if the `new_core` branch gets merged before this one.

Reply:
I mean why not use `task.get_status()` or `task.get_due_date()` instead of going through getattr?

Reply:
It's just a way to have a declarative way to set field translator. Having a getter allows some overriding of the `Field` class.